Update EIP-2935: update gas cost & implementation, and specify verkle_fork requirements (#8573)
ethereum/EIPs: EIPS/eip-2935.md

## Specification

| Parameter                 | Value                                        |
| ------------------------- | -------------------------------------------- |
| `FORK_TIMESTAMP`          | TBD                                          |
| `VERKLE_FORK_TIMESTAMP`   | TBD                                          |
| `HISTORY_STORAGE_ADDRESS` | `0x25a219378dad9b3503c8268c9ca836a52427a4fb` |
| `HISTORY_SERVE_WINDOW`    | `8192`                                       |

> | `BLOCKHASH_OLD_WINDOW`    | `256`                                        |

Reviewer: ```suggestion
| `BLOCKHASH_SERVE_WINDOW`    | `256`                                        |
```
I think this name would be better since we wouldn't modify blockhash window even when we serve it from state

This EIP specifies for storing last `HISTORY_SERVE_WINDOW` block hashes in a ring buffer storage of `HISTORY_SERVE_WINDOW` length.

At the start of processing any block where `block.timestamp >= FORK_TIMESTAMP` (ie. before processing any transactions), begin storing the blockhashes in the following way:

```python
def process_block_hash_history(block: Block, state: State):
    if block.timestamp >= FORK_TIMESTAMP:
        state.insert_slot(HISTORY_STORAGE_ADDRESS, (block.number-1) % HISTORY_SERVE_WINDOW , block.parent.hash)
```

> Note that from the `fork_block` (first block where `block.timestamp >= FORK_TIMESTAMP`), the implementation of `BLOCKHASH` does not change. The blockhashes of the last `BLOCKHASH_OLD_WINDOW` blocks can be retrieved as they previously were. `BLOCKHASH` will return `0` for blocks preceding the `BLOCKHASH_OLD_WINDOW`.

Reviewer: ```suggestion
Note that from the `fork_block` (first block where `block.timestamp >= FORK_TIMESTAMP`), the implementation of `BLOCKHASH` does not change. The blockhashes of the last `BLOCKHASH_SERVE_WINDOW` blocks can be retrieved as they previously were. `BLOCKHASH` will continue to return `0` for blocks preceding the `BLOCKHASH_SERVE_WINDOW`.
```

> Blockhashes from the `HISTORY_SERVE_WINDOW` can be retrieved directly from storage for blocks whose `block.number + 1 >= fork_block`.

Reviewer 2: at fork block, the parent is inserted, so it's actually `block.number >= fork_block -1`

Author: `block.number + 1 >= fork_block` is equivalent to `block.number >= fork_block - 1`, no?

Reviewer 2: 🤦 yes

> Blockhashes from the `HISTORY_SERVE_WINDOW` can be retrieved directly from storage for blocks whose `block.number + 1 >= fork_block`.

Reviewer 3: Can this be clarified? I think the idea is that if you want a block within `HISTORY_SERVE_WINDOW` but larger than `BLOCKHASH_OLD_WINDOW` you now call into the contract - right? Can this sentence then be updated such that it is clear that the contract should now be called and thus that `BLOCKHASH` will still put 0 on stack (even though the data is in the contract storage)

  return state.load_slot(HISTORY_STORAGE_ADDRESS, arg % HISTORY_SERVE_WINDOW)
```

> <!-- TODO: Do we need to update the implementation because there is not history backfilling? -->

Reviewer 2: no, because this is only really a problem one verkle activates (presumably people will keep using `BLOCKHASH` until the buffer is filled enough that it becomes interesting to use the contract. I would, however, add a note to explain just what i said.

Reviewer: also blockhash only serves its original 256 window right?

### Contract Implementation

> Exact evm assembly that can be used for the contract to resolve `BLOCKHASH` from verkle_block_fork

Reviewer 2: that's not correct: the contract is available before the prague fork, whether it resolves blockhash or not.

Reviewer 2: so I would just remove the reference to `BLOCKHASH`

```

> // check if input > 8 byte value and revert if this isn't the case

Reviewer 2: ideally, the contract would be updated to add a `set` function

  "v": "0x1b",
  "r": "0x539",
  "s": "0x1b9b6eb1f0",
  "hash": "7ba81426bfa88a2cf4ea5c9abbbe83619505acd1173bc8450f93cf17cde3784b"
}
```

Note, the input in the transaction has a simple constructor prefixing the desired runtime code.

The sender of the transaction can be calculated as `0xa4690f0ed0d089faa1e0ad94c8f1b4a2fd4b0734`. The address of the first contract deployed from the account is `rlp([sender, 0])` which equals `0x25a219378dad9b3503c8268c9ca836a52427a4fb`. This is how `HISTORY_STORAGE_ADDRESS` is determined. Although this style of contract creation is not tied to any specific initcode like create2 is, the synthetic address is cryptographically bound to the input data of the transaction (e.g. the initcode).

Some fork (blocks where `block.timestamp >= FORK_TIMESTAMP`) activation scenarios:

- For the fork to be activated at genesis, no history is written to the genesis state, and at the start of block `1`, genesis hash will be written as a normal operation to slot `0`.

> - for fork activation at block `1`, only genesis hash will be written at slot `0`.

Reviewer 4: Nit: maybe a good moment to uppercase this and the bullets below?

> - for fork activation at block `32`, block `31`'s hash will be written to slot `31`. Additional history for block `0..30` hashes will continue to be served, but will not be backfilled into storage.

Reviewer 4: ```suggestion
- for fork activation at block `32`, block `31`'s hash will be written to slot `31`. Additional history for block `0..30` hashes will continue to be served by `BLOCKHASH`, but will not be backfilled into the contract storage.
```

> - for fork activation at block `10000`, block `9999`'s hash will be written to slot `9999`. Additional history for the the `BLOCKHASH_OLD_WINDOW` (`9744..9998`) will continue to be served, but will not be backfilled into storage. History for the blocks preceding the `HISTORY_SERVE_WINDOW` (`1808..9998`) will not be served, and will not be backfilled into storage. At block `20000`, the BLOCKHASH opcode would continue serving history for the `BLOCKHASH_OLD_WINDOW` (`19744..19999`) and would return 0 for blocks before the `BLOCKHASH_OLD_WINDOW`. The blocks from `HISTORY_SERVE_WINDOW` would be served from state for blocks `11808..19999`. Blocks prior to the `HISTORY_SERVE_WINDOW` would return 0.

Reviewer 2: this should be broken into 2 examples and "will not be served" should be replaced by "will return 0" for clarity

> - for fork activation at block `10000`, block `9999`'s hash will be written to slot `9999`. Additional history for the the `BLOCKHASH_OLD_WINDOW` (`9744..9998`) will continue to be served, but will not be backfilled into storage. History for the blocks preceding the `HISTORY_SERVE_WINDOW` (`1808..9998`) will not be served, and will not be backfilled into storage. At block `20000`, the BLOCKHASH opcode would continue serving history for the `BLOCKHASH_OLD_WINDOW` (`19744..19999`) and would return 0 for blocks before the `BLOCKHASH_OLD_WINDOW`. The blocks from `HISTORY_SERVE_WINDOW` would be served from state for blocks `11808..19999`. Blocks prior to the `HISTORY_SERVE_WINDOW` would return 0.

Reviewer 3: The slot is incorrect, it should be `9999 % HISTORY_SERVE_WINDOW`.

Reviewer 3: Can it be made explicit also that the "served from state" is thus served from state via the contract call and thus NOT via `BLOCKHASH` opcode?

For verkle-fork (blocks where `block.timestamp >= VERKLE_FORK_TIMESTAMP`), we consider that the fork (`FORK_TIMESTAMP`) is already in effect for at least `HISTORY_SERVE_WINDOW` blocks (or from genesis if `verkle_fork_block < HISTORY_SERVE_WINDOW`). The following changes take effect:

- `BLOCKHASH` MUST retrieve blockhashes from the contract storage
- The gas cost of the `BLOCKHASH` opcode is updated as per specification
- The storage access is provided in the verkle execution witness.

### [EIP-158](./eip-158.md) handling

This address is currently exempt from [EIP-158](./eip-158.md) cleanup in Kaustinen Verkle Testnet but we plan to address this in the following way:

> - Deploy a contract à la [EIP-4788](./eip-4788.md) which just supports `get` method to resolve the BLOCKHASH as per the logic defined in `resolve_blockhash` (and use the generated address as the BLOCKHASH contract address).

Reviewer 2: should have a `set` as well

Author: Is this for 2935, or for the verkle one?

Second concern was how to best transition the BLOCKHASH resolution logic post fork by:

1. Either waiting for `HISTORY_SERVE_WINDOW` blocks for the entire relevant history to persist
2. Storing of all last `HISTORY_SERVE_WINDOW` block hashes on the fork block.

<!-- TODO: Explain why we are choosing the former -->

## Backwards Compatibility

> The behavior of `BLOCKHASH` opcode gets extended in a backward compatible manner as its functionality, and its gas cost, remains the same from `FORK_TIMESTAMP`. The blockhashes from `BLOCKHASH_OLD_WINDOW` can still be retrieve with `BLOCKHASH`. The additional history can be directly retrieve from the contract storage. From `VERKLE_FORK_TIMESTAMP`, `BLOCKHASH` will serve history for all blocks from `HISTORY_SERVE_WINDOW` from the contract storage address at the updated gas cost.

Author: Ok, thanks for pointing that out, I had misunderstood and thought that the BLOCKHASH opcode was eventually updated to serve HISTORY_SERVE_WINDOW. Upcoming EIP will be adjusted.
